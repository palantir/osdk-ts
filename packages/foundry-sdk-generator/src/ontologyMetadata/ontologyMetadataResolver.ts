/*
 * Copyright 2023 Palantir Technologies, Inc. All rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import type { components } from "@osdk/gateway";
import { namespaces } from "@osdk/gateway";
import type { WireOntologyDefinition } from "@osdk/generator";
import { UserTokenAuth } from "@osdk/legacy-client";
import { getApiRequestFunction } from "../net/FetchClient.js";
import { Result } from "./Result.js";

export class OntologyMetadataResolver {
  #authToken: string;
  constructor(authToken: string, private stackName: string) {
    this.#authToken = authToken;
  }

  private getRequestFunction<T>() {
    return getApiRequestFunction<T>(
      new UserTokenAuth({ userToken: this.#authToken }),
      this.stackName,
      `foundry-typescript-osdk-generator/${process.env.npm_package_version!}`,
      "/api",
    );
  }

  private filterMetadata(
    ontologyFullMetadata: components.OntologyFullMetadata,
    expectedEntities: {
      linkTypes: Map<string, Set<string>>;
      objectTypes: Set<string>;
      queryTypes: Set<string>;
      actionTypes: Set<string>;
      interfaceTypes: Set<string>;
    },
  ): components.OntologyFullMetadata {
    const filteredObjectTypes = Object.fromEntries(
      Object.entries(ontologyFullMetadata.objectTypes).filter((
        [objectTypeApiName],
      ) => expectedEntities.objectTypes.has(objectTypeApiName.toLowerCase())),
    );

    const filteredInterfaceTypes = Object.fromEntries(
      Object.entries(ontologyFullMetadata.interfaceTypes).filter((
        [interfaceApiName],
      ) => expectedEntities.interfaceTypes.has(interfaceApiName.toLowerCase())),
    );

    Object.values(filteredObjectTypes).forEach(objectType => {
      const linkTypesToKeep = expectedEntities.linkTypes.get(
        objectType.objectType.apiName.toLowerCase(),
      );
      if (!linkTypesToKeep) {
        objectType.linkTypes = [];
        return;
      }

      objectType.linkTypes = objectType.linkTypes.filter(linkType =>
        linkTypesToKeep.has(linkType.apiName.toLowerCase())
      );
    });

    const filteredActionTypes = Object.fromEntries(
      Object.entries(ontologyFullMetadata.actionTypes).filter(
        ([actionApiName]) => {
          if (
            expectedEntities.actionTypes.has(
              this.camelize(actionApiName).toLowerCase(),
            )
          ) {
            return true;
          }
          return false;
        },
      ),
    );

    const filteredQueryTypes = Object.fromEntries(
      Object.entries(ontologyFullMetadata.queryTypes).filter(([queryApiName]) =>
        expectedEntities.queryTypes.has(queryApiName.toLowerCase())
      ),
    );

    return {
      ontology: ontologyFullMetadata.ontology,
      objectTypes: filteredObjectTypes,
      actionTypes: filteredActionTypes,
      queryTypes: filteredQueryTypes,
      interfaceTypes: filteredInterfaceTypes,
      sharedPropertyTypes: {},
    };
  }

  public async getWireOntologyDefinition(
    ontologyRid: string,
    entities: {
      actionTypesApiNamesToLoad?: string[];
      objectTypesApiNamesToLoad?: string[];
      queryTypesApiNamesToLoad?: string[];
      interfaceTypesApiNamesToLoad?: string[];
      linkTypesApiNamesToLoad?: string[];
    },
  ): Promise<Result<WireOntologyDefinition, string[]>> {
    let ontology: components.Ontology;
    try {
      ontology = await namespaces.getOntology(
        this.getRequestFunction(),
        ontologyRid,
      );
    } catch (e) {
      return Result.err([
        `Unable to load the specified Ontology with network error: ${
          JSON.stringify(e)
        }`,
      ]);
    }

    const ontologyFullMetadata = await namespaces.getOntologyFullMetadata<
      components.OntologyFullMetadata
    >(
      this.getRequestFunction(),
      ontology.rid,
    );

    if ((ontologyFullMetadata as any).errorName != null) {
      return Result.err([
        `Unable to load the specified Ontology metadata.\n${
          JSON.stringify(ontologyFullMetadata, null, 2)
        }`,
      ]);
    }

    const linkTypes = new Map<string, Set<string>>();
    const objectTypes = new Set(
      entities.objectTypesApiNamesToLoad?.map(object => object.toLowerCase()),
    );
    const queryTypes = new Set(
      entities.queryTypesApiNamesToLoad?.map(query => query.toLowerCase()),
    );
    const actionTypes = new Set(
      entities.actionTypesApiNamesToLoad?.map(action =>
        this.camelize(action).toLowerCase()
      ),
    );

    const interfaceTypes = new Set(
      entities.interfaceTypesApiNamesToLoad?.map(interfaceName =>
        interfaceName.toLowerCase()
      ),
    );

    for (const linkType of entities.linkTypesApiNamesToLoad ?? []) {
      const [objectTypeApiName, linkTypeApiName] = linkType.toLowerCase().split(
        ".",
      );
      if (!linkTypes.has(objectTypeApiName)) {
        linkTypes.set(objectTypeApiName, new Set());
      }
      linkTypes.get(objectTypeApiName)?.add(linkTypeApiName);
    }

    const filteredFullMetadata = this.filterMetadata(ontologyFullMetadata, {
      objectTypes,
      linkTypes,
      actionTypes,
      queryTypes,
      interfaceTypes,
    });

    const validData: Result<{}, string[]> = this.validateLoadedOntologyMetadata(
      filteredFullMetadata,
      {
        objectTypes,
        linkTypes,
        actionTypes,
        queryTypes,
      },
    );

    if (validData.isErr()) {
      return Result.err(validData.error) as Result<
        WireOntologyDefinition,
        string[]
      >;
    }
    return Result.ok(filteredFullMetadata) as Result<
      WireOntologyDefinition,
      string[]
    >;
  }

  private validateLoadedOntologyMetadata(
    filteredFullMetadata: components.OntologyFullMetadata,
    expectedEntities: {
      linkTypes: Map<string, Set<string>>;
      objectTypes: Set<string>;
      queryTypes: Set<string>;
      actionTypes: Set<string>;
    },
  ): Result<{}, string[]> {
    const errors: string[] = [];
    const loadedObjectTypes = Object.fromEntries(
      Object.values(filteredFullMetadata.objectTypes).map(object => [
        object.objectType.apiName.toLowerCase(),
        object,
      ]),
    );

    const loadedLinkTypes = Object.fromEntries(
      Object.values(filteredFullMetadata.objectTypes).map(object => [
        object.objectType.apiName.toLowerCase(),
        Object.fromEntries(
          object.linkTypes.map(link => [link.apiName.toLowerCase(), link]),
        ),
      ]),
    );

    const missingObjectTypes: string[] = [];
    for (const object of expectedEntities.objectTypes) {
      // Expected an object that wasn't loaded
      if (!loadedObjectTypes[object]) {
        missingObjectTypes.push(object);
        continue;
      }

      // Expected a link that wasn't received
      for (const expectedLink of expectedEntities.linkTypes.get(object) ?? []) {
        if (!loadedLinkTypes[object][expectedLink]) {
          errors.push(
            `Unable to find link type ${expectedLink} for Object Type ${object}`,
          );
        }
      }

      for (const [, link] of Object.entries(loadedLinkTypes[object])) {
        // Loaded a link where target was not loaded
        if (
          !expectedEntities.objectTypes.has(
            link.objectTypeApiName.toLowerCase(),
          )
        ) {
          errors.push(
            `Unable to load link type ${link.apiName} for ${
              loadedObjectTypes[object].objectType.apiName
            }, because the target object type ${link.objectTypeApiName} is not loaded. Please specify the target Object type with --objectTypes ${link.objectTypeApiName}`,
          );
        }
      }
    }

    if (missingObjectTypes.length > 0) {
      errors.push(
        `Unable to find the following Object Types: ${missingObjectTypes.join()}`,
      );
    }

    const loadedQueryTypes = Object.fromEntries(
      Object.entries(filteredFullMetadata.queryTypes).map((
        [queryApiName, query],
      ) => [
        queryApiName.toLowerCase(),
        query,
      ]),
    );

    // Validate expected queries were loaded
    const missingQueryTypes: string[] = [];
    for (const queryApiName of expectedEntities.queryTypes) {
      if (!loadedQueryTypes[queryApiName]) {
        missingQueryTypes.push(queryApiName);
      }
    }

    // Validate parameters and output for Queries
    for (const query of Object.values(loadedQueryTypes)) {
      const result = this.validateQueryParametersAndOutput(
        query,
        expectedEntities.objectTypes,
      );
      if (result.isErr()) {
        for (const errorString of result.error) {
          errors.push(errorString);
        }
      }
    }

    if (missingQueryTypes.length > 0) {
      errors.push(
        `Unable to find the following Query Types: ${missingQueryTypes.join()}`,
      );
    }

    const loadedActionTypes = Object.fromEntries(
      Object.entries(filteredFullMetadata.actionTypes).map((
        [actionApiName, action],
      ) => [
        this.camelize(actionApiName).toLowerCase(),
        action,
      ]),
    );

    const missingActionTypes: string[] = [];
    for (const actionApiName of expectedEntities.actionTypes) {
      if (!loadedActionTypes[actionApiName]) {
        missingActionTypes.push(actionApiName);
      }
    }

    // Validate parameters for Actions
    for (const action of Object.values(loadedActionTypes)) {
      const result = this.validateActionParameters(
        action,
        expectedEntities.objectTypes,
      );
      if (result.isErr()) {
        for (const errorString of result.error) {
          errors.push(errorString);
        }
      }
    }

    if (missingActionTypes.length > 0) {
      errors.push(
        `Unable to find the following Action Types: ${missingActionTypes.join()}`,
      );
    }

    if (errors.length > 0) {
      return Result.err(errors);
    }
    return Result.ok({});
  }

  private validateQueryParametersAndOutput(
    query: components.QueryTypeV2,
    loadedObjectApiNames: Set<string>,
  ): Result<{}, string[]> {
    const parameterValidation: Array<Result<{}, string[]>> = Object.entries(
      query.parameters,
    ).map(
      ([paramName, paramData]) =>
        this.visitSupportedQueryTypes(
          query.apiName,
          paramName,
          paramData.dataType,
          loadedObjectApiNames,
        ),
    );

    parameterValidation.push(
      this.visitSupportedQueryTypes(
        query.apiName,
        "output",
        query.output,
        loadedObjectApiNames,
      ),
    );

    const results = Result.coalesce<{}, string>(parameterValidation);

    return results;
  }

  private validateActionParameters(
    actionType: components.ActionTypeV2,
    loadedObjectApiNames: Set<string>,
  ): Result<{}, string[]> {
    const camelizedApiName = this.camelize(actionType.apiName);

    const parameterValidation: Array<Result<{}, string[]>> = Object.entries(
      actionType.parameters,
    ).map(
      ([_paramName, paramData]) =>
        this.isSupportedActionTypeParameter(
          camelizedApiName,
          paramData.dataType,
          loadedObjectApiNames,
        ),
    );

    return Result.coalesce<{}, string>(parameterValidation);
  }

  private visitSupportedQueryTypes(
    queryApiName: string,
    propertyName: string,
    baseType: components.QueryDataType,
    loadedObjectApiNames: Set<string>,
  ): Result<{}, string[]> {
    switch (baseType.type) {
      case "array":
      case "set":
        return this.visitSupportedQueryTypes(
          queryApiName,
          propertyName,
          baseType.subType,
          loadedObjectApiNames,
        );
      case "objectSet":
      case "object":
        if (
          loadedObjectApiNames.has(baseType.objectTypeApiName?.toLowerCase()!)
        ) {
          return Result.ok({});
        }
        return Result.err([
          `Unable to load query ${queryApiName} because it takes an unloaded object type as a parameter: ${baseType
            .objectTypeApiName!} in parameter ${propertyName}. `
          + `Make sure to specify it as an argument with --ontologyObjects ${baseType
            .objectTypeApiName!}.}`,
        ]);
      case "struct":
        const results = baseType.fields?.map(field => {
          return this.visitSupportedQueryTypes(
            queryApiName,
            propertyName,
            field.fieldType,
            loadedObjectApiNames,
          );
        });

        if (!results) {
          return Result.ok({});
        }

        return Result.coalesce(results);
      case "union":
        if (
          baseType.unionTypes.find(unionTypes => unionTypes.type === "null")
        ) {
          return Result.ok({});
        }
        return Result.err([
          `Unable to load query ${queryApiName} because it takes an unsupported parameter type: ${
            JSON.stringify(
              baseType,
            )
          } in parameter ${propertyName}`,
        ]);
      case "string":
      case "boolean":
      case "attachment":
      case "date":
      case "double":
      case "float":
      case "integer":
      case "long":
      case "threeDimensionalAggregation":
      case "timestamp":
      case "twoDimensionalAggregation":
      case "null":
        return Result.ok({});
      case "unsupported":
      default:
        return Result.err([
          `Unable to load query ${queryApiName} because it takes an unsupported parameter type: ${
            JSON.stringify(
              baseType,
            )
          } in parameter ${propertyName}`,
        ]);
    }
  }

  private isSupportedActionTypeParameter(
    actionApiName: string,
    actonTypeParameter: components.ActionParameterType,
    loadedObjectApiNames: Set<string>,
  ): Result<{}, string[]> {
    switch (actonTypeParameter.type) {
      case "array":
        return this.isSupportedActionTypeParameter(
          actionApiName,
          actonTypeParameter.subType,
          loadedObjectApiNames,
        );
      case "object":
        if (
          loadedObjectApiNames.has(
            actonTypeParameter.objectTypeApiName?.toLowerCase()!,
          )
        ) {
          return Result.ok({});
        }
        return Result.err([
          `Unable to load action ${actionApiName} because it takes an unloaded object type as a parameter: ${actonTypeParameter
            .objectTypeApiName!} `
          + `make sure to specify it as an argument with --ontologyObjects ${actonTypeParameter
            .objectTypeApiName!})`,
        ]);
      case "objectSet":
        if (
          loadedObjectApiNames.has(
            actonTypeParameter.objectTypeApiName?.toLowerCase()!,
          )
        ) {
          return Result.ok({});
        }
        return Result.err([
          `Unable to load action ${actionApiName} because it takes an ObjectSet of unloaded object type as a parameter: ${actonTypeParameter
            .objectTypeApiName!} `
          + `make sure to specify it as an argument with --ontologyObjects ${actonTypeParameter
            .objectTypeApiName!})`,
        ]);
      case "string":
      case "boolean":
      case "attachment":
      case "date":
      case "double":
      case "integer":
      case "long":
      case "timestamp":
        return Result.ok({});
      default:
        return Result.err([
          `Unable to load action ${actionApiName} because it takes an unsupported parameter: ${
            JSON.stringify(
              actonTypeParameter,
            )
          } `
          + `specify only the actions you want to load with the --actions argument.`,
        ]);
    }
  }

  private camelize(name: string) {
    return name.replace(/-./g, segment => segment[1]!.toUpperCase());
  }
}
