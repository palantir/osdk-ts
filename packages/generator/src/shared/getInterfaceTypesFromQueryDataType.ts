/*
 * Copyright 2025 Palantir Technologies, Inc. All rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import type { QueryDataType } from "@osdk/foundry.ontologies";

export function getInterfaceTypesFromQueryDataType(
  dataType: QueryDataType,
  types: Set<string>,
): void {
  switch (dataType.type) {
    case "array":
    case "set":
      getInterfaceTypesFromQueryDataType(dataType.subType, types);
      return;
    case "interfaceObject":
    case "interfaceObjectSet":
      types.add(dataType.interfaceTypeApiName!);
      return;

    case "struct":
      for (const prop of Object.values(dataType.fields)) {
        getInterfaceTypesFromQueryDataType(prop.fieldType, types);
      }
      return;

    case "union":
      for (const type of dataType.unionTypes) {
        getInterfaceTypesFromQueryDataType(type, types);
      }
      return;

    case "entrySet":
      getInterfaceTypesFromQueryDataType(dataType.keyType, types);
      getInterfaceTypesFromQueryDataType(dataType.valueType, types);
      return;

    case "attachment":
    case "boolean":
    case "date":
    case "double":
    case "float":
    case "integer":
    case "long":
    case "null":
    case "string":
    case "threeDimensionalAggregation":
    case "timestamp":
    case "twoDimensionalAggregation":
    case "entrySet":
    case "unsupported":
    case "object":
    case "objectSet":
      /* complete no-op */
      return;

    default:
      const _: never = dataType;
      throw new Error(
        `Unknown query data type ${(dataType as any).type}`,
      );
  }
}
