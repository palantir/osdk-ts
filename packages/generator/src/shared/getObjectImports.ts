/*
 * Copyright 2024 Palantir Technologies, Inc. All rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import type { EnhancedObjectType } from "../GenerateContext/EnhancedObjectType.js";
import type { ForeignType } from "../GenerateContext/ForeignType.js";

export function getObjectImports(
  objects: Set<EnhancedObjectType | ForeignType>,
  curApiName: string | undefined,
  currentFilePath: string,
  v2: boolean,
) {
  return Array.from(objects).filter(obj => obj.fullApiName !== curApiName)
    .map(obj => {
      return `import type { ${obj.getDefinitionIdentifier(v2)} as ${
        obj.getImportedDefinitionIdentifier(v2)
      } } from "${obj.getImportPathRelTo("./" + currentFilePath)}";`;
    }).join("\n");
}
