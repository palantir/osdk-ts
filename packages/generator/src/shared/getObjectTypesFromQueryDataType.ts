/*
 * Copyright 2024 Palantir Technologies, Inc. All rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import type { QueryDataType } from "@osdk/gateway/types";

export function getObjectTypesFromQueryDataType(
  dataType: QueryDataType,
  types: Set<string>,
) {
  switch (dataType.type) {
    case "array":
    case "set":
      getObjectTypesFromQueryDataType(dataType.subType, types);
      return;

    case "object":
      types.add(dataType.objectTypeApiName);
      return;

    case "objectSet":
      types.add(dataType.objectTypeApiName!);
      return;

    case "struct":
      for (const prop of Object.values(dataType.fields)) {
        getObjectTypesFromQueryDataType(prop.fieldType, types);
      }
      return;

    case "union":
      for (const type of dataType.unionTypes) {
        getObjectTypesFromQueryDataType(type, types);
      }
      return;

    case "attachment":
    case "boolean":
    case "date":
    case "double":
    case "float":
    case "integer":
    case "long":
    case "null":
    case "string":
    case "threeDimensionalAggregation":
    case "timestamp":
    case "twoDimensionalAggregation":
    case "unsupported":
      /* complete no-op */
      return;

    default:
      const _: never = dataType;
      throw new Error(
        `Cannot find object types from unsupported QueryDataType ${
          (dataType as any).type
        }`,
      );
  }
}
