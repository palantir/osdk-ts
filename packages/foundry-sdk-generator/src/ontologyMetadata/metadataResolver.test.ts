/*
 * Copyright 2023 Palantir Technologies, Inc. All rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { apiServer, handlers } from "@osdk/shared.test";
import { afterAll, beforeAll, describe, expect, it } from "vitest";
import { OntologyMetadataResolver } from "./ontologyMetadataResolver.js";

describe("Load Ontologies Metadata", () => {
  beforeAll(() => {
    apiServer.listen();
  });

  afterAll(() => {
    apiServer.close();
  });

  it("Loads no object types and action types", async () => {
    const ontologyMetadataResolver = new OntologyMetadataResolver(
      "myAccessToken",
      "https://stack.palantir.com",
    );
    const ontologyDefinitions = await ontologyMetadataResolver
      .getWireOntologyDefinition(
        "ri.ontology.main.ontology.698267cc-6b48-4d98-beff-29beb24e9361",
        {
          objectTypesApiNamesToLoad: [],
          actionTypesApiNamesToLoad: [],
          queryTypesApiNamesToLoad: [],
        },
      );

    if (ontologyDefinitions.isErr()) {
      throw new Error();
    }

    expect(ontologyDefinitions.value).toMatchInlineSnapshot(`
      {
        "actionTypes": {},
        "interfaceTypes": {},
        "objectTypes": {},
        "ontology": {
          "apiName": "default-ontology",
          "description": "The default ontology",
          "displayName": "Ontology",
          "rid": "ri.ontology.main.ontology.698267cc-6b48-4d98-beff-29beb24e9361",
        },
        "queryTypes": {},
        "sharedPropertyTypes": {},
      }
    `);
  });

  it("Fails on missing entities for wire ontology", async () => {
    const ontologyMetadataResolver = new OntologyMetadataResolver(
      "myAccessToken",
      "https://stack.palantir.com",
    );
    const ontologyDefinitions = await ontologyMetadataResolver
      .getWireOntologyDefinition(
        "ri.ontology.main.ontology.698267cc-6b48-4d98-beff-29beb24e9361",
        {
          objectTypesApiNamesToLoad: ["objectDoesNotExist", "employee"],
          actionTypesApiNamesToLoad: ["action-does-not-exit"],
          linkTypesApiNamesToLoad: ["employee.doesNotExist"],
          queryTypesApiNamesToLoad: ["queryDoesNotExist"],
        },
      );

    if (ontologyDefinitions.isOk()) {
      throw new Error();
    }

    expect(ontologyDefinitions.error).toMatchInlineSnapshot(`
      [
        "Unable to find link type doesnotexist for Object Type employee",
        "Unable to find the following Object Types: objectdoesnotexist",
        "Unable to find the following Query Types: querydoesnotexist",
        "Unable to find the following Action Types: actiondoesnotexit",
      ]
    `);
  });

  it("Captures errors while loading wire ontology definition", async () => {
    const ontologyMetadataResolver = new OntologyMetadataResolver(
      "myAccessToken",
      "https://stack.palantir.com",
    );
    apiServer.use(...handlers.unsupportedMetadataHandler);
    const ontologyDefinitions = await ontologyMetadataResolver
      .getWireOntologyDefinition(
        "ri.ontology.main.ontology.698267cc-6b48-4d98-beff-29beb24e9361",
        {
          objectTypesApiNamesToLoad: undefined,
          actionTypesApiNamesToLoad: ["unsupportedAction"],
          queryTypesApiNamesToLoad: ["returnsObject"],
        },
      );

    if (ontologyDefinitions.isOk()) {
      throw new Error();
    }

    expect(ontologyDefinitions.error).toMatchInlineSnapshot(`
      [
        "Unable to load query returnsObject because it takes an unloaded object type as a parameter: Employee in parameter output. Make sure to specify it as an argument with --ontologyObjects Employee.}",
        "Unable to load action unsupportedAction because it takes an unsupported parameter: {"type":"unsupportedType"} specify only the actions you want to load with the --actions argument.",
        "Unable to load action unsupportedAction because it takes an unloaded object type as a parameter: unsupported make sure to specify it as an argument with --ontologyObjects unsupported)",
      ]
    `);
  });

  it("Does not load any entity type using generatePackageV2", async () => {
    const ontologyMetadataResolver = new OntologyMetadataResolver(
      "myAccessToken",
      "https://stack.palantir.com",
    );
    const ontologyDefinitions = await ontologyMetadataResolver
      .getWireOntologyDefinition(
        "ri.ontology.main.ontology.698267cc-6b48-4d98-beff-29beb24e9361",
        {
          objectTypesApiNamesToLoad: undefined,
          actionTypesApiNamesToLoad: undefined,
          linkTypesApiNamesToLoad: undefined,
        },
      );

    if (ontologyDefinitions.isErr()) {
      throw new Error();
    }
    expect(Object.keys(ontologyDefinitions.value.objectTypes)).toHaveLength(0);
    expect(Object.keys(ontologyDefinitions.value.actionTypes)).toHaveLength(0);
    expect(Object.keys(ontologyDefinitions.value.queryTypes)).toHaveLength(0);
  });

  it("Loads object and action types without link types", async () => {
    const ontologyMetadataResolver = new OntologyMetadataResolver(
      "myAccessToken",
      "https://stack.palantir.com",
    );
    const ontologyDefinitions = await ontologyMetadataResolver
      .getWireOntologyDefinition(
        "ri.ontology.main.ontology.698267cc-6b48-4d98-beff-29beb24e9361",
        {
          objectTypesApiNamesToLoad: ["employee"],
          actionTypesApiNamesToLoad: ["promote-employee"],
        },
      );

    if (ontologyDefinitions.isErr()) {
      throw new Error();
    }

    expect(Object.keys(ontologyDefinitions.value.objectTypes)).toHaveLength(1);
    expect(ontologyDefinitions.value.objectTypes.Employee.linkTypes)
      .toHaveLength(0);
    expect(Object.keys(ontologyDefinitions.value.actionTypes)).toHaveLength(1);
    expect(ontologyDefinitions.value).toMatchSnapshot();
  });

  it("Loads object and action types using only specified link types", async () => {
    const ontologyMetadataResolver = new OntologyMetadataResolver(
      "myAccessToken",
      "https://stack.palantir.com",
    );
    const ontologyDefinitions = await ontologyMetadataResolver
      .getWireOntologyDefinition(
        "ri.ontology.main.ontology.698267cc-6b48-4d98-beff-29beb24e9361",
        {
          objectTypesApiNamesToLoad: ["Employee", "Office"],
          actionTypesApiNamesToLoad: ["promote-employee"],
          linkTypesApiNamesToLoad: ["Employee.peeps", "Employee.officeLink"],
        },
      );

    if (ontologyDefinitions.isErr()) {
      throw new Error(ontologyDefinitions.error.join("\n"));
    }

    expect(Object.keys(ontologyDefinitions.value.objectTypes)).toHaveLength(2);
    expect(ontologyDefinitions.value.objectTypes.Employee.linkTypes)
      .toHaveLength(2);
    expect(ontologyDefinitions.value.objectTypes.Office.linkTypes).toHaveLength(
      0,
    );
    expect(Object.keys(ontologyDefinitions.value.actionTypes)).toHaveLength(1);
    expect(ontologyDefinitions.value).toMatchSnapshot();
  });
});
