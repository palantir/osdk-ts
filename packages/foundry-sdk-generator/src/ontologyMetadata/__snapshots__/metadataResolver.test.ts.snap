// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Load Ontologies Metadata > Loads object and action types using only specified link types 1`] = `
{
  "actionTypes": {
    "promote-employee": {
      "apiName": "promote-employee",
      "description": "Update an employee's title and compensation",
      "operations": [
        {
          "objectTypeApiName": "Employee",
          "type": "modifyObject",
        },
      ],
      "parameters": {
        "employeeId": {
          "dataType": {
            "type": "integer",
          },
          "required": true,
        },
        "newCompensation": {
          "dataType": {
            "type": "double",
          },
          "required": true,
        },
        "newTitle": {
          "dataType": {
            "type": "string",
          },
          "required": true,
        },
      },
      "rid": "ri.ontology.main.action-type.7ed72754-7491-428a-bb18-4d7296eb2167",
      "status": "ACTIVE",
    },
  },
  "interfaceTypes": {},
  "objectTypes": {
    "Employee": {
      "implementsInterfaces": [
        "FooInterface",
      ],
      "implementsInterfaces2": {
        "FooInterface": {
          "properties": {
            "fooSpt": "fullName",
          },
        },
      },
      "linkTypes": [
        {
          "apiName": "peeps",
          "cardinality": "MANY",
          "displayName": "Peeps",
          "objectTypeApiName": "Employee",
          "status": "EXPERIMENTAL",
        },
        {
          "apiName": "officeLink",
          "cardinality": "ONE",
          "displayName": "Office",
          "objectTypeApiName": "Office",
          "status": "EXPERIMENTAL",
        },
      ],
      "objectType": {
        "apiName": "Employee",
        "description": "A full-time or part-time 

 employee of our firm",
        "primaryKey": "employeeId",
        "properties": {
          "employeeId": {
            "dataType": {
              "type": "integer",
            },
          },
          "employeeStatus": {
            "dataType": {
              "itemType": {
                "type": "string",
              },
              "type": "timeseries",
            },
            "description": "TimeSeries of the status of the employee",
          },
          "fullName": {
            "dataType": {
              "type": "string",
            },
          },
          "office": {
            "dataType": {
              "type": "string",
            },
            "description": "The unique "ID" of the employee's \\"primary\\" assigned office.
 This is some more text.",
          },
          "startDate": {
            "dataType": {
              "type": "date",
            },
            "description": "The date the employee was hired (most recently, if they were re-hired)",
          },
        },
        "rid": "ri.ontology.main.object-type.401ac022-89eb-4591-8b7e-0a912b9efb44",
        "status": "ACTIVE",
        "titleProperty": "fullName",
      },
      "sharedPropertyTypeMapping": {
        "fooSpt": "fullName",
      },
    },
    "Office": {
      "implementsInterfaces": [],
      "implementsInterfaces2": {},
      "linkTypes": [],
      "objectType": {
        "apiName": "Office",
        "description": "A office in our Company",
        "primaryKey": "officeId",
        "properties": {
          "entrance": {
            "dataType": {
              "type": "geopoint",
            },
          },
          "name": {
            "dataType": {
              "type": "string",
            },
            "description": "The Name of the Office",
          },
          "occupiedArea": {
            "dataType": {
              "type": "geoshape",
            },
            "description": "The occupied area of the Office",
          },
          "officeId": {
            "dataType": {
              "type": "string",
            },
          },
        },
        "rid": "ri.ontology.main.object-type.404ac022-89eb-4591-8b7e-1a912b9efb45",
        "status": "ACTIVE",
        "titleProperty": "officeId",
      },
      "sharedPropertyTypeMapping": {},
    },
  },
  "ontology": {
    "apiName": "default-ontology",
    "description": "The default ontology",
    "displayName": "Ontology",
    "rid": "ri.ontology.main.ontology.698267cc-6b48-4d98-beff-29beb24e9361",
  },
  "queryTypes": {},
  "sharedPropertyTypes": {},
}
`;

exports[`Load Ontologies Metadata > Loads object and action types without link types 1`] = `
{
  "actionTypes": {
    "promote-employee": {
      "apiName": "promote-employee",
      "description": "Update an employee's title and compensation",
      "operations": [
        {
          "objectTypeApiName": "Employee",
          "type": "modifyObject",
        },
      ],
      "parameters": {
        "employeeId": {
          "dataType": {
            "type": "integer",
          },
          "required": true,
        },
        "newCompensation": {
          "dataType": {
            "type": "double",
          },
          "required": true,
        },
        "newTitle": {
          "dataType": {
            "type": "string",
          },
          "required": true,
        },
      },
      "rid": "ri.ontology.main.action-type.7ed72754-7491-428a-bb18-4d7296eb2167",
      "status": "ACTIVE",
    },
  },
  "interfaceTypes": {},
  "objectTypes": {
    "Employee": {
      "implementsInterfaces": [
        "FooInterface",
      ],
      "implementsInterfaces2": {
        "FooInterface": {
          "properties": {
            "fooSpt": "fullName",
          },
        },
      },
      "linkTypes": [],
      "objectType": {
        "apiName": "Employee",
        "description": "A full-time or part-time 

 employee of our firm",
        "primaryKey": "employeeId",
        "properties": {
          "employeeId": {
            "dataType": {
              "type": "integer",
            },
          },
          "employeeStatus": {
            "dataType": {
              "itemType": {
                "type": "string",
              },
              "type": "timeseries",
            },
            "description": "TimeSeries of the status of the employee",
          },
          "fullName": {
            "dataType": {
              "type": "string",
            },
          },
          "office": {
            "dataType": {
              "type": "string",
            },
            "description": "The unique "ID" of the employee's \\"primary\\" assigned office.
 This is some more text.",
          },
          "startDate": {
            "dataType": {
              "type": "date",
            },
            "description": "The date the employee was hired (most recently, if they were re-hired)",
          },
        },
        "rid": "ri.ontology.main.object-type.401ac022-89eb-4591-8b7e-0a912b9efb44",
        "status": "ACTIVE",
        "titleProperty": "fullName",
      },
      "sharedPropertyTypeMapping": {
        "fooSpt": "fullName",
      },
    },
  },
  "ontology": {
    "apiName": "default-ontology",
    "description": "The default ontology",
    "displayName": "Ontology",
    "rid": "ri.ontology.main.ontology.698267cc-6b48-4d98-beff-29beb24e9361",
  },
  "queryTypes": {},
  "sharedPropertyTypes": {},
}
`;
